### PR TITLE
Chore: Migrate lib/encryption folder to TypeScript

### PR DESCRIPTION
**Pull request by gerzonc**
## Issue(s)	

Migrate `encryption` folder to TypeScript

## How to test or reproduce

E2E Encryption should be fully tested

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/lib/encryption/encryption.ts [hunk 1]
```diff
@@ -271,22 +287,22 @@ class Encryption {
...
-				sub =>
+				(sub: ISubscription) =>
 					// Encrypted message
 					sub?.lastMessage?.t === E2E_MESSAGE_TYPE &&
 					// Message pending decrypt
 					sub?.lastMessage?.e2e === E2E_STATUS.PENDING
 			);
 			await Promise.all(
-				subsToDecrypt.map(async sub => {
+				subsToDecrypt.map(async (sub: any) => {
```
Comment (AlexAlexandre) [inline app/lib/encryption/encryption.ts:297]: Can we remove this any ? maybe `TSubscriptionModel` ?

File: app/lib/encryption/encryption.ts [hunk 2]
```diff
@@ -340,12 +356,12 @@ class Encryption {
...
 
 		try {
-			const batch = [];
+			const batch: (Model | null | void | false | Promise<void>)[] = [];
 			// If the subscription doesn't exists yet
 			if (!subRecord) {
 				// Let's create the subscription with the data received
 				batch.push(
-					subCollection.prepareCreate(s => {
+					subCollection.prepareCreate((s: any) => {
```
Comment (AlexAlexandre) [inline app/lib/encryption/encryption.ts:364]: same thing here, mabye `ISubscription`

File: app/lib/encryption/encryption.ts [hunk 3]
```diff
@@ -355,7 +371,7 @@ class Encryption {
 				if (!subRecord._hasPendingUpdate) {
 					// Let's update the subscription with the received E2EKey
 					batch.push(
-						subRecord.prepareUpdate(s => {
+						subRecord.prepareUpdate((s: any) => {
```
Comment (AlexAlexandre) [inline app/lib/encryption/encryption.ts:374]: same thing here

**Review by AlexAlexandre — CHANGES_REQUESTED** [review]

File: app/lib/encryption/utils.ts [hunk 4]
```diff
@@ -33,28 +33,29 @@ export const bufferToB64URI = buffer => {
 	return base64;
 };
 // SimpleCrypto.utils.convertArrayBufferToUtf8 is not working with unicode emoji
-export const bufferToUtf8 = buffer => {
+export const bufferToUtf8 = (buffer: ArrayBuffer): string => {
 	const uintArray = new Uint8Array(buffer);
-	const encodedString = String.fromCharCode.apply(null, uintArray);
+	const encodedString = String.fromCharCode.apply(null, uintArray as unknown as number[]);
```
[the remark this answers is not in the document]
Comment (gerzonc) [reply app/lib/encryption/utils.ts:38]: Did better using `const uintArray = new Uint8Array(buffer) as number[] & Uint8Array`

File: app/lib/encryption/utils.ts [hunk 5]
```diff
@@ -33,28 +33,29 @@ export const bufferToB64URI = buffer => {
...
 	const decodedString = decodeURIComponent(escape(encodedString));
 	return decodedString;
```
[the remark this answers is not in the document]
Comment (gerzonc) [reply app/lib/encryption/utils.ts:40]: True :)

File: app/lib/encryption/encryption.ts (hunk 1 above)
Comment (gerzonc) [reply app/lib/encryption/encryption.ts:297]: Correct

File: app/lib/encryption/room.ts [hunk 6]
```diff
@@ -19,9 +21,19 @@ import {
...
+import { IUser } from '../../definitions/IUser';
 
 export default class EncryptionRoom {
-	constructor(roomId, userId) {
+	ready: boolean;
+	roomId: string;
+	userId: string;
+	establishing: boolean;
+	readyPromise: Deferred<unknown>;
+	sessionKeyExportedString!: string | ByteBuffer;
```
Comment (GleidsonDaniel) [inline app/lib/encryption/room.ts:32]: No need to use Non-null assertion.

File: app/lib/encryption/room.ts [hunk 7]
```diff
@@ -19,9 +21,19 @@ import {
...
+	keyID!: string;
```
Comment (GleidsonDaniel) [inline app/lib/encryption/room.ts:33]: You are using Non-null assertion because it was not declared, wouldn't it be better to declare it?

**Review by GleidsonDaniel — CHANGES_REQUESTED** [review]

File: app/definitions/IMessage.ts [hunk 8]
```diff
@@ -54,8 +59,10 @@ export interface ILastMessage {
 }
 
 export interface IMessage {
+	_id: string;
+	rid: string;
 	msg?: string;
-	t?: SubscriptionType;
+	t?: MessageType;
```
Comment (gerzonc) [inline app/definitions/IMessage.ts:65]: This change is done between IMessage, IThread and IThreadMessage; not actually sure about it, but `toDecrypt` complains a lot with this.

Comment (AlexAlexandre) [reply app/definitions/IMessage.ts:65]: I'm not sure about, the difference between `SubscriptionType` and `MessageType` is huge.
Can you debug and try find out if is ok change this ?

File: app/lib/encryption/room.ts [hunk 9]
```diff
@@ -207,8 +222,8 @@ export default class EncryptionRoom {
 	};
 
 	// Decrypt text
-	decryptText = async msg => {
-		msg = b64ToBuffer(msg.slice(12));
+	decryptText = async (msg: string | ArrayBuffer): Promise<any> => {
```
Comment (AlexAlexandre) [inline app/lib/encryption/room.ts:225]: You can change this `Promise<any>` to `Promise<string>`

**Review by AlexAlexandre — CHANGES_REQUESTED** [review]

**Review by AlexAlexandre — APPROVED** [review]
LGMT

**Pull request merged**